packages/next/src/client/form.tsx

```diff
@@ -0,0 +1,326 @@
+import { useMergedRef } from './use-merged-ref'
+import type { AppRouterInstance } from '../shared/lib/app-router-context.shared-runtime'
+
+const DISALLOWED_FORM_PROPS = ['method', 'encType', 'target'] as const
+
+type HTMLFormProps = HTMLProps<HTMLFormElement>
+type DisallowedFormProps = (typeof DISALLOWED_FORM_PROPS)[number]
+
+export type FormProps = {
+  /**
+   * `action` can be either a `string` or a function.
+   * - If `action` is a string, it will be interpreted as a path or URL to navigate to when the form is submitted.
+   *   The path will be prefetched when the form becomes visible.
+   * - If `action` is a function, it will be called when the form is submitted. See the [React docs](https://react.dev/reference/react-dom/components/form#props) for more.
+   */
+  action: NonNullable<HTMLFormProps['action']>
+  /**
+   * Whether submitting the form should replace the current `history` state instead of adding a new url into the stack.
+   * Only valid if `action` is a string.
+   *
+   * @defaultValue `false`
+   */
+  replace?: boolean
+  /**
+   * Override the default scroll behavior when navigating.
+   * Only valid if `action` is a string.
+   *
+   * @defaultValue `true`
+   */
+  scroll?: boolean
+} & Omit<HTMLFormProps, 'action' | DisallowedFormProps>
+
+export default function Form({
+  replace,
+  scroll,
+  ref: externalRef,
+  ...props
+}: FormProps) {
+  const actionProp = props.action
+
+  for (const key of DISALLOWED_FORM_PROPS) {
+    if (key in props) {
+      if (process.env.NODE_ENV === 'development') {
+        console.error(
+          `<Form> does not support changing \`${key}\`. ` +
+            (typeof actionProp === 'string'
+              ? `If you'd like to use it to perform a mutation, consider making \`action\` a function instead.\n` +
+                `Learn more: https://nextjs.org/docs/app/building-your-application/data-fetching/server-actions-and-mutations`
+              : '')
+        )
+      }
+      delete (props as Record<string, unknown>)[key]
+    }
+  }
+
+  const router = useRouter()
+
+  const [setIntersectionRef, isVisible] = useIntersection({
+    rootMargin: '200px',
+    disabled: typeof actionProp !== 'string', // if we don't have an action path, we can't preload anything anyway.
+  })
+
+  // TODO: Link does something like this in addition, do we need it?
+  // check if visible state need to be reset
+  // if (previousActionProp.current !== actionProp) {
+  //   resetVisible()
+  //   previousActionProp.current = actionProp
+  // }
+  const ownRef = useMergedRef<HTMLFormElement>(
+    setIntersectionRef,
+    externalRef ?? null
+  )
+
+  useEffect(() => {
+    if (typeof actionProp !== 'string') {
+      return
+    }
+
+    if (!isVisible) {
+      return
+    }
+
+    try {
+      // TODO: do we need to take the current field values here?
+      // or are we assuming that queryparams can't affect this (but what about rewrites)?
+      router.prefetch(actionProp)
```

Comment:
do we want to specify an "auto" prefetch type here? I believe by default this will be a "full" prefetch which means it'll do the dynamic request rather than just the loading.js

Reply:
omg yeah definitely, good catch!